torch/_inductor/optimize_indexing.py

```diff
@@ -136,77 +156,105 @@ def constant(value, dtype):
 
     @staticmethod
     def reciprocal(x):
-        return ValueRanges.checked_unary_map(x, lambda y: 1 / y)
+        x = ValueRanges.wrap(x)
+        if 0 in x:
+            return ValueRanges(-math.inf, math.inf)
+        else:
+            return ValueRanges.decreasing_map(x, lambda y: 1 / y)
+
+    @staticmethod
+    def square(x):
+        return ValueRanges.convex_min_zero_map(x, lambda y: y * y)
 
     @staticmethod
     def abs(x):
-        return ValueRanges.checked_unary_map(x, abs)
+        return ValueRanges.convex_min_zero_map(x, abs)
 
     @staticmethod
     def neg(x):
-        return ValueRanges.checked_unary_map(x, lambda x: -x)
+        return ValueRanges.decreasing_map(x, operator.neg)
 
     @staticmethod
     def truediv(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.truediv)
+        b = ValueRanges.wrap(b)
+        if 0 in b:
+            return ValueRanges(-math.inf, math.inf)
+        else:
+            return ValueRangeAnalysis.mul(a, ValueRanges(1 / b.upper, 1 / b.lower))
 
     @staticmethod
     def div(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.truediv)
+        # We think of this as floor(a / b)
+        out = ValueRangeAnalysis.truediv(a, b)
+        floor = sympy.functions.elementary.integers.floor
+        return ValueRanges(floor(out.lower), floor(out.upper))
 
     @staticmethod
     def add(a, b):
-        return ValueRanges.binary_map(a, b, operator.add)
+        return ValueRanges.coordinatewise_increasing_map(a, b, operator.add)
 
     @staticmethod
     def mul(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.mul)
+        return ValueRanges.coordinatewise_monotone_map(a, b, operator.mul)
 
     @staticmethod
     def sub(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.sub)
+        b = ValueRanges.wrap(b)
+        return ValueRangeAnalysis.add(a, ValueRanges(-b.upper, -b.lower))
 
     @staticmethod
     def exp(x):
-        return ValueRanges.unary_map(x, sympy.functions.elementary.exponential.exp)
-
-    @staticmethod
-    def square(x):
-        return ValueRanges.checked_unary_map(x, lambda y: y * y)
+        return ValueRanges.increasing_map(x, sympy.functions.elementary.exponential.exp)
 
     @staticmethod
     def log(x):
-        return ValueRanges.checked_unary_map(
+        return ValueRanges.increasing_map(
             x, lambda y: -math.inf if y <= 0 else sympy.log(y)
         )
 
     @staticmethod
     def sqrt(x):
-        return ValueRanges.unary_map(x, sympy.sqrt)
+        return ValueRanges.increasing_map(x, sympy.sqrt)
 
     @staticmethod
     def pow(a, b):
-        return ValueRanges.binary_map_products(a, b, operator.pow)
+        def is_integer(val):
+            return isinstance(val, int) or (
```

Comment:
we might want to update the `to_dtype` to accurately reflect dtype conversions if we are depending on it.

Reply:
Where are we depending on `to_dtype`?

Reply:
we're not depending on `to_dtype`, but we are depending on whether the a value is an int or float.